sycli: improved filter syntax

evanrichter opened the pull request:
this is my first try at #180, and everything seems to be working well!

The only thing I'm not sure of is how you want to do error handling on clearly invalid strings like "peers>five". This would currently panic with the message "Invalid numeric value"

Commit:
sycli: improved filter syntax using regex

fixes #180

Luminarys commented on sycli/src/main.rs:
```diff
@@ -665,3 +667,120 @@ fn main() {
+                let field = search.split(delim).nth(0).unwrap().to_string();
+                let value = search.split(delim).nth(1).unwrap();
+
+                let value = if field == "name"
```
I think it's a good idea to do a match here and be explicit about possible field names we could expect. Then we can print to stderr that an unrecognized field name was used if nothing matches.

Luminarys commented on sycli/src/main.rs:
```diff
@@ -665,3 +667,120 @@ fn main() {
+    // possible delimiters
+    let delimiters = vec![">=", "<=", "==", "!=", "<", ">", ":"];
+
+    'search: for search in searches.split_whitespace() {
```
I wonder if there's a convenient crate to do more advanced tokenization. receptor supports things like name:"spaced name", but it seems like this wouldn't work so well here. If it's very inconvenient to find a good looking library for this though I think this is good enough for now.

Luminarys commented on sycli/src/main.rs:
```diff
@@ -665,3 +667,120 @@ fn main() {
+    }
+
+    // if no delimiters found, assume a simple name query
+    if criterion.is_empty() {
```
Can you add a test which has a valid filter condition, then invalid, e.g. "name:foo key~val" to show this branch doesn't happen per loop iteration.

Luminarys replied:
Also can you add a test which shows that a filter like "spaced name" gets translated into this condition with its whitespace

Luminarys reviewed (commented):
Thanks for doing this!

evanrichter commented:
@Luminarys ok I think I addressed everything! At a high level, I'm using the regex crate now, building two regexes (one for string types and one for numeric), and now either can be quoted or unquoted, and just like before if nothing matches right, the string is put into a name:<input> search. Also I added a delimiter `::` for case-sensitive searches. `:` is still case insensitive

again, I'm glad to contribute and open to suggestions

Luminarys commented:
Looks good now, thanks!
